File: sdks/java/src/main/java/com/apollodeploy/signal/sdk/api/SegmentsAPI.java
package com.apollodeploy.signal.sdk.api;

import com.fasterxml.jackson.core.type.TypeReference;
import com.apollodeploy.signal.sdk.models.Types.*;
import com.apollodeploy.signal.sdk.transport.Transport;
import java.time.Duration;
import java.util.List;

/** SegmentsAPI API operations. */
public final class SegmentsAPI {
    private final Transport transport;

    public SegmentsAPI(Transport transport) {
        this.transport = transport;
    }

    public SegmentPageResponse listSegments(
        String projectId
    ) {
        String path = "/v1/projects/${projectId}/segments";
        path = path.replace("${projectId}", Transport.encodePathSegment(String.valueOf(projectId)));

        Transport.Request request = new Transport.Request("GET", path);
        return transport.execute(request, new TypeReference<SegmentPageResponse>() {
        });
    }

    public SegmentResponse getSegment(
        String projectId,
        String segmentId
    ) {
        String path = "/v1/projects/${projectId}/segments/${segmentId}";
        path = path.replace("${projectId}", Transport.encodePathSegment(String.valueOf(projectId)));
        path = path.replace("${segmentId}", Transport.encodePathSegment(String.valueOf(segmentId)));

        Transport.Request request = new Transport.Request("GET", path);
        return transport.execute(request, new TypeReference<SegmentResponse>() {
        });
    }

    public CursorPage listContactsInSegment(
        String projectId,
        String segmentId
    ) {
        String path = "/v1/projects/${projectId}/segments/${segmentId}/contacts";
        path = path.replace("${projectId}", Transport.encodePathSegment(String.valueOf(projectId)));
        path = path.replace("${segmentId}", Transport.encodePathSegment(String.valueOf(segmentId)));

        Transport.Request request = new Transport.Request("GET", path);
        return transport.execute(request, new TypeReference<CursorPage>() {
        });
    }

    public SegmentResponse createSegment(
        String projectId,
        CreateSegmentBody body
    ) {
        String path = "/v1/projects/${projectId}/segments";
        path = path.replace("${projectId}", Transport.encodePathSegment(String.valueOf(projectId)));

        Transport.Request request = new Transport.Request("POST", path);
        request.body(body, "application/json");
        return transport.execute(request, new TypeReference<SegmentResponse>() {
        });
    }

    public void deleteSegment(
        String projectId,
        String segmentId
    ) {
        String path = "/v1/projects/${projectId}/segments/${segmentId}";
        path = path.replace("${projectId}", Transport.encodePathSegment(String.valueOf(projectId)));
        path = path.replace("${segmentId}", Transport.encodePathSegment(String.valueOf(segmentId)));

        Transport.Request request = new Transport.Request("DELETE", path);
        transport.executeVoid(request);
    }

}

File: sdks/kotlin/src/main/kotlin/com/apollodeploy/signal/sdk/api/SegmentsAPI.kt
package com.apollodeploy.signal.sdk.api

import com.apollodeploy.signal.sdk.internal.Transport
import com.apollodeploy.signal.sdk.models.*
import io.ktor.http.HttpMethod
import io.ktor.http.encodeURLPathPart
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.coroutines.flow.Flow

/** SegmentsAPI API operations. */
class SegmentsAPI(
    private val transport: Transport,
) {
/** listSegments operation. */
    suspend fun listSegments(
        projectId: String,
    ): SegmentPageResponse {
        var path = "/v1/projects/\${projectId}/segments"
        path = path.replace("\${projectId}", projectId.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        return transport.execute<SegmentPageResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** getSegment operation. */
    suspend fun getSegment(
        projectId: String,
        segmentId: String,
    ): SegmentResponse {
        var path = "/v1/projects/\${projectId}/segments/\${segmentId}"
        path = path.replace("\${projectId}", projectId.toString().encodeURLPathPart())
        path = path.replace("\${segmentId}", segmentId.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        return transport.execute<SegmentResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** listContactsInSegment operation. */
    suspend fun listContactsInSegment(
        projectId: String,
        segmentId: String,
    ): CursorPage {
        var path = "/v1/projects/\${projectId}/segments/\${segmentId}/contacts"
        path = path.replace("\${projectId}", projectId.toString().encodeURLPathPart())
        path = path.replace("\${segmentId}", segmentId.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        return transport.execute<CursorPage>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** createSegment operation. */
    suspend fun createSegment(
        projectId: String,
        body: CreateSegmentBody,
    ): SegmentResponse {
        var path = "/v1/projects/\${projectId}/segments"
        path = path.replace("\${projectId}", projectId.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        return transport.execute<SegmentResponse>(
            method = HttpMethod.Post,
            path = path,
            query = query.ifEmpty { null },
            body = Json.encodeToJsonElement(body),
            contentType = "application/json",
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** deleteSegment operation. */
    suspend fun deleteSegment(
        projectId: String,
        segmentId: String,
    ): Unit {
        var path = "/v1/projects/\${projectId}/segments/\${segmentId}"
        path = path.replace("\${projectId}", projectId.toString().encodeURLPathPart())
        path = path.replace("\${segmentId}", segmentId.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        transport.executeRaw(
            method = HttpMethod.Delete,
            path = path,
            query = query.ifEmpty { null },
            body = JsonObject(emptyMap()),
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
        return Unit
    }

}
